File: src/robot_racing_car/nodes/steering_controller.py
#!/usr/bin/env python

import rospy
import Adafruit_PCA9685

from std_msgs.msg import UInt8

class SteeringController:
    LEFT_ANGLE = 1
    RIGHT_ANGLE = -1

    LEFT_PULSE = 460
    RIGHT_PULSE = 260

    def steering_callback(self, message):
        self.update(message.data)

    def map_range(x, X_min, X_max, Y_min, Y_max):
        # Linear mapping between two ranges of values
        X_range = X_max - X_min
        Y_range = Y_max - Y_min
        XY_ratio = X_range / Y_range

        y = ((x - X_min) / XY_ratio + Y_min) // 1

        return int(y)

    def update(self, angle):
        pulse = self.map_range(angle, self.RIGHT_ANGLE, self.LEFT_ANGLE, self.RIGHT_PULSE, self.LEFT_PULSE)

        self.steering_actuator.set_pwm(self.actuator_channel, 0, pulse)

    def __init__(self):
        self.steering_actuator = Adafruit_PCA9685.PCA9685()
        self.steering_actuator.set_pwm_freq(60)
        self.actuator_channel = 1

        rospy.Subscriber("robot_car_steering", UInt8, self.steering_callback)

if __name__ == "__main__":
    rospy.init_node("steering_controller", log_level=rospy.DEBUG)

    controller = SteeringController()

    rospy.spin()

File: src/steering_controller/src/SteeringController.cpp
#include <ros/ros.h>
#include <sensor_msgs/Joy.h>
#include "std_msgs/Float32.h"
#include "PCA9685.h"

const int LEFT_ANGLE = 1;
const int RIGHT_ANGLE = -1;

const int LEFT_PULSE = 460;
const int RIGHT_PULSE = 260;

PCA9685 *actuator;

int getActuatorPulseValue(float joystickValue, int joystickMin, int joystickMax, int actuatorMin, int actuatorMax)
{
	// Linear mapping between two ranges of values
	
    int joystickRange = joystickMax - joystickMin;
    int actuatorRange = actuatorMax - actuatorMin;
    
    float joystickActuatorRatio = (float)joystickRange / (float)actuatorRange;
    int actuatorValue = (joystickValue - joystickMin) / joystickActuatorRatio + actuatorMin;

	return actuatorValue;
}

void steeringCallback(const sensor_msgs::Joy::ConstPtr& joy)
{
	int pulse = getActuatorPulseValue(joy->axes[0], RIGHT_ANGLE, LEFT_ANGLE, RIGHT_PULSE, LEFT_PULSE);
	
	actuator->setPWM(1, pulse);
	
	ROS_INFO("Steering value: [%i]", pulse);
}

int main(int argc, char* argv[])
{ 
	ros::init(argc, argv, "steering_controller");

	ros::NodeHandle nodeHandle;

	actuator = new PCA9685(1, 64);
	actuator->setPWMFreq(60);
  
	ros::Subscriber subscriber = nodeHandle.subscribe<sensor_msgs::Joy>("joy", 1, steeringCallback);
	
	ROS_INFO("Steering controller ROS node started.");

	ros::spin();
	
	return 0;
}
